src/lib_rules.ml

```diff
@@ -120,6 +120,36 @@ module Gen (P : Install_rules.Params) = struct
       ~sandbox:alias_module_build_sandbox
       ~dep_graphs:(Ocamldep.Dep_graphs.dummy m)
 
+  let build_deprecated_modules (lib : Library.t)
+        cctx
+        ~modules
+        ~js_of_ocaml
+        ~dynlink
+        ~(deprecated : Module.t Module.Name.Map.t) =
+    let lib_name = String.capitalize (Lib_name.Local.to_string lib.name) in
+    let transition_message =
+      match lib.wrapped with
+      | Simple _ -> "" (* will never be accessed anyway *)
+      | Yes_with_transition r -> r
+    in
+    Module.Name.Map.iteri deprecated ~f:(fun name m ->
+      let contents =
+        let name = Module.Name.to_string name in
+        let hidden_name = sprintf "%s__%s" lib_name name in
+        let real_name = sprintf "%s.%s" lib_name name in
+        sprintf {|[@@@deprecated "%s. Use %s instead."] include %s|}
+          transition_message real_name hidden_name
+      in
+      let source_path = Option.value_exn (Module.file m Impl) in
+      Build.return contents
+      >>> Build.write_file_dyn source_path
+      |> SC.add_rule sctx
+    );
+    let dep_graphs =
+      Ocamldep.Dep_graphs.deprecated ~modules ~deprecated in
+    let cctx = Compilation_context.set_modules cctx deprecated in
```

Comment:
We should use a dedicated compilation context here, i.e. with preprocessing, special flags, libraries, etc... Since these modules are under our control, we know exactly how they should be compiled. It should be similar to the one for the alias module